datafusion/expr/src/type_coercion/binary.rs

```diff
@@ -66,194 +193,68 @@ fn mathematics_temporal_result_type(
-        Operator::StringConcat => string_concat_coercion(lhs_type, rhs_type),
-    };
-
-    // re-write the error message of failed coercions to include the operator's information
-    result.ok_or(DataFusionError::Plan(format!("{lhs_type:?} {op} {rhs_type:?} can't be evaluated because there isn't a common type to coerce the types to")))
+    rhs: &DataType,
+) -> Result<(DataType, DataType)> {
+    signature(lhs, op, rhs).map(|sig| (sig.lhs, sig.rhs))
 }
 
 /// Coercion rules for mathematics operators between decimal and non-decimal types.
-pub fn math_decimal_coercion(
+fn math_decimal_coercion(
     lhs_type: &DataType,
     rhs_type: &DataType,
-) -> (Option<DataType>, Option<DataType>) {
+) -> Option<(DataType, DataType)> {
     use arrow::datatypes::DataType::*;
 
-    if both_decimal(lhs_type, rhs_type) {
-        return (None, None);
-    }
-
     match (lhs_type, rhs_type) {
-        (Null, dec_type @ Decimal128(_, _)) => (Some(dec_type.clone()), None),
-        (dec_type @ Decimal128(_, _), Null) => (None, Some(dec_type.clone())),
         (Dictionary(key_type, value_type), _) => {
-            let (value_type, rhs_type) = math_decimal_coercion(value_type, rhs_type);
-            let lhs_type = value_type
-                .map(|value_type| Dictionary(key_type.clone(), Box::new(value_type)));
-            (lhs_type, rhs_type)
+            let (value_type, rhs_type) = math_decimal_coercion(value_type, rhs_type)?;
+            Some((Dictionary(key_type.clone(), Box::new(value_type)), rhs_type))
         }
         (_, Dictionary(key_type, value_type)) => {
-            let (lhs_type, value_type) = math_decimal_coercion(lhs_type, value_type);
-            let rhs_type = value_type
-                .map(|value_type| Dictionary(key_type.clone(), Box::new(value_type)));
-            (lhs_type, rhs_type)
+            let (lhs_type, value_type) = math_decimal_coercion(lhs_type, value_type)?;
+            Some((lhs_type, Dictionary(key_type.clone(), Box::new(value_type))))
         }
-        (Decimal128(_, _), Float32 | Float64) => (Some(Float64), Some(Float64)),
-        (Float32 | Float64, Decimal128(_, _)) => (Some(Float64), Some(Float64)),
-        (Decimal128(_, _), _) => {
-            let converted_decimal_type = coerce_numeric_type_to_decimal(rhs_type);
-            (None, converted_decimal_type)
+        (Null, dec_type @ Decimal128(_, _)) | (dec_type @ Decimal128(_, _), Null) => {
+            Some((dec_type.clone(), dec_type.clone()))
         }
-        (_, Decimal128(_, _)) => {
-            let converted_decimal_type = coerce_numeric_type_to_decimal(lhs_type);
-            (converted_decimal_type, None)
+        (Decimal128(_, _), Decimal128(_, _)) => {
+            Some((lhs_type.clone(), rhs_type.clone()))
         }
-        _ => (None, None),
+        // Unlike with comparison we don't coerce to a decimal in the case of floating point
+        // numbers, instead falling back to floating point arithmetic instead
```

Comment:
I'm not sure if this behaviour was intentional or accidental, but I opted to preserve it